frontend/lib/src/components/widgets/DateInput/DateInput.tsx

```diff
@@ -74,360 +65,320 @@ function datesToStrings(dates: Date[]): string[] {
-      this.updateFromProtobuf()
-    } else {
-      this.commitWidgetValue({ fromUi: false })
-    }
-  }
-
-  public componentDidUpdate(): void {
-    this.maybeUpdateFromProtobuf()
-  }
-
-  public componentWillUnmount(): void {
-    this.formClearHelper.disconnect()
-  }
-
-  private maybeUpdateFromProtobuf(): void {
-    const { setValue } = this.props.element
-    if (setValue) {
-      this.updateFromProtobuf()
-    }
-  }
-
-  private updateFromProtobuf(): void {
-    const { value: values } = this.props.element
-    this.props.element.setValue = false
-    const dateValues = values.map((v: string) => new Date(v))
-    this.setState(
-      {
-        values: dateValues,
-        isEmpty: !dateValues,
-      },
-      () => {
-        this.commitWidgetValue({ fromUi: false })
-      }
-    )
-  }
-
-  /** Commit state.value to the WidgetStateManager. */
-  private commitWidgetValue = (source: Source): void => {
-    const { widgetMgr, element, fragmentId } = this.props
-    widgetMgr.setStringArrayValue(
-      element,
-      datesToStrings(this.state.values),
-      source,
-      fragmentId
-    )
-  }
-
+function DateInput({
+  disabled,
+  element,
+  widgetMgr,
+  width,
+  fragmentId,
+}: Props): ReactElement {
   /**
-   * If we're part of a clear_on_submit form, this will be called when our
-   * form is submitted. Restore our default value and update the WidgetManager.
+   * An array with start and end date specified by the user via the UI. If the user
+   * didn't touch this widget's UI, the default value is used. End date is optional.
    */
-  private onFormCleared = (): void => {
-    const defaultValue = stringsToDates(this.props.element.default)
-    this.setState(
-      {
-        values: defaultValue,
-        isEmpty: !defaultValue,
-      },
-      () => this.commitWidgetValue({ fromUi: true })
-    )
-  }
-
-  private handleChange = ({
-    date,
-  }: {
-    date: Date | (Date | null | undefined)[] | null | undefined
-  }): void => {
-    if (date === null || date === undefined) {
-      this.setState({
-        values: [],
-        isEmpty: true,
-      })
-      return
-    }
-
-    const values: Date[] = []
-    if (Array.isArray(date)) {
-      date.forEach((dt: Date | null | undefined) => {
-        if (dt) {
-          values.push(dt)
-        }
-      })
-    } else {
-      values.push(date)
-    }
-
-    this.setState(
-      {
-        values,
-        isEmpty: !values,
-      },
-      () => {
-        if (!this.state.isEmpty) this.commitWidgetValue({ fromUi: true })
+  const [value, setValueWithSource] = useBasicWidgetState<
+    Date[],
+    DateInputProto
+  >({
+    getStateFromWidgetMgr,
+    getDefaultStateFromProto,
+    getCurrStateFromProto,
+    updateWidgetMgrState,
+    element,
+    widgetMgr,
+    fragmentId,
+  })
+
+  const [isEmpty, setIsEmpty] = useState(false)
+
+  const theme = useTheme()
+  const { colors, fontSizes, lineHeights, spacing } = theme
```

Comment:
super nit: we could unpack it in one line: `const { colors, fontSizes, lineHeights, spacing } = useTheme()` and also add `sizes` here and do not re-use `theme.spacing`/`theme.sizes` further below and thus do not mix it with the destructured access